docs/upgrade.md

```diff
@@ -0,0 +1,51 @@
+# Upgrade
+
+Currently the upgrade can be performed just between follow up downstream builds
+e.g. 171 -> 172.
+There is opened [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1767400) which
+tracking the issue which prevents us to upgrade to the latest available build.
+
+There is prepared config option `upgrade_to_latest` in the DEPLOYMENT section
+which can be set or will be the default value when the BZ gets resolved.
+
+## Deployment
+
+To be able upgrade your cluster you cannot deploy the latest downstream
+build, because then you won't have any next build available for upgrade.
+
+For this purpose we have this config file: [upgrade.yaml](/conf/ocsci/upgrade.yaml)
+which you can pass via `--ocsci-conf` when you are running the deployment. This will
+automatically take not latest build but one before.
+
+## Upgrade execution
+
+For upgrade we have few pytest marks defined [here](/ocs_ci/framework/pytest_customization/marks.py):
+
+* **before_upgrade** - mark tests which are suppose to run before upgrade
+* **upgrade** - this is mark for upgrade itself
+* **after_upgrade** - mark tests which are suppose to run after upgrade
+
+Those marks has to be imported from mentioned
+[modul](/ocs_ci/framework/pytest_customization/marks.py) and your test have to
+be decorated by those marks as we combine those mark decorators with order
+decorator used by `pytest-ordering` plugin to properly order the tests.
+
+To execute the upgrade you can use this command:
+
+```bash
+run-ci tests/
+    --cluster-name kerberos_ID-ocs-deployment \
+    --cluster-path /home/my_user/my-ocs-dir \
+    -m 'before_upgrade or upgrade or after_upgrade'
+ ```
+
+Of course if you would like to run just the upgrade and not any other tests
+before and after upgrade you can use just `-m upgrade`
+
+## TODO
+
+Once there will be 4.3 build and we will have different branch for OCS-CI we
```

Comment:
As said in previous comment, I dont see the usecase for separate branch or before and after_upgrade marks, It adds complexity.

Reply:
Once we will have 4.3. 4.3 will be master For this repo and for 4.2 we will have separate branch for sure.

Reply:
Are you talking about ocs or ocs-ci here?

Reply:
yes, as the product will evolve, we will have changes in deployment, will have new tests and so on which will be written for latest version which means that to keep everything aligned with version we will have master branch of ocs-ci for the latest supported version we have (for example 4.3 once we will have). For previous versions we will have separate branch and if some change will be needed to be backported to 4.2 for example we will cherry-pick those specific changes but development of new tests will be done just for latest version.

Or how else you would like to solve all those changes between versions?

Reply:
So we have a couple issues here regarding versioning of the product (using 4.2 and 4.3 as examples):
- Test cases that do not apply to both 4.2 and 4.3
- Fixes that will need to be back ported to 4.2 branch
- Deployment logic that differs from 4.2 to 4.3

These problems can be solved using the following (without a separate branch):
- Tags (or module?) for tests that do not apply to the old version, 4.2 in our example. Or we can use the inverse and have a tag that denotes tests that _can_ run on the old version. We can do whatever is easiest.
- Conditionals in our deployment code to change between 4.2 and 4.3, hook into config to determine which we are deploying. Difficulty depends on how much changes between versions.
- Fixes do not need to be back ported if we remain on one branch.

In my opinion, I think we should stay clear of creating a separate branch if at all possible. I think maintaining another branch is more trouble than it will ultimately be worth. However, this entirely depends on how much goes into the deployment code to ensure it works for both versions, how many test cases will not be applicable to the old version, and how many fixes we plan to back port into the 'old' branch.

Reply:
We should definitely discuss in the meeting before going forward with assumption that we need 2 branches to test upgrades, The *right* thing is to discuss what the upgrade usecases are and why *branching* is required to *achieve* that.  Branching is a costly operation and we cannot put a put a strain on users to maintain multiple branches.

Reply:
As Coady suggested there are multiple ways to achieve upgrade logic in single branch, Infact just folder separate eg: upgrade_4.2, upgrade_4.3, can also be added to above bullet points for specific tests taht needs separation.

Reply:
resolved in meeting.